[inductor] Slightly faster memory allocation on CPU

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118255
* __->__ #118171
* #118070
* #118065

Based on `python benchmarks/dynamo/microbenchmarks/overheads.py`:
- Before `12.2us`
- After `10.5us`

This is inspired by https://github.com/pytorch/pytorch/commit/a2c17a2b00f7c41866bbde28d33b8c50e5632e01 -- but in Python rather than C++


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

jansel committed:
[inductor] Slightly faster memory allocation on CPU

[ghstack-poisoned]

jgong5 reviewed (approved):
Do we have test to cover the allocation of >8D tensors?

peterbell10 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1161,6 +1162,16 @@ def make_buffer_allocation(self, buffer):
     def make_allocation(self, name, device, dtype, shape, stride):
+        if device.type == "cpu" and len(shape) <= 8:
+            # optimized path for faster allocations, saving ~2us versus the stuff below
+            # TODO(jansel): can we do similar things on the GPU?
```
I also wrote `empty_strided_cuda` in `ATen/cuda/EmptyTensor.h` though you would have to deal with cpu-only builds where it may not be available. Possibly through a dispatch stub, or `ifdef USE_CUDA` may be enough.

You could also use `empty_cpu` and `empty_cuda` for the contiguous cases. This should be slightly faster still as they are able to do fast `TensorImpl` restriding, though after python overhead I'm not sure how much that will amount to.

jansel replied:
Next PR in the stack will add `empty_strided_cuda`.

peterbell10 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -585,12 +585,56 @@ static PyObject* assert_size_stride(PyObject* dummy, PyObject* args) {
+    return THPVariable_Wrap(at::detail::empty_strided_cpu(
+        at::IntArrayRef(sizes, ndim1), at::IntArrayRef(strides, ndim2), dtype));
+  } catch (std::exception const& ex) {
+    PyErr_SetString(PyExc_RuntimeError, ex.what());
```
Use `HANDLE_TH_ERRORS`/`END_HANDLE_TH_ERRORS` for proper exception translation.

jansel commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -585,12 +585,56 @@ static PyObject* assert_size_stride(PyObject* dummy, PyObject* args) {
+    TORCH_CHECK(PyTuple_GET_SIZE(args) == 3);
+
+    // note PyTuple_GET_ITEM returns a borrowed ref, so no need for refcounts
+    ssize_t sizes[8];
```
It was easy enough to route >8 to `at::empty_strided` -- they don't happen in practice.

jansel replied:
Fixed now

jansel committed:
Update on "[inductor] Slightly faster memory allocation on CPU"


Based on `python benchmarks/dynamo/microbenchmarks/overheads.py`:
- Before `12.2us`
- After `10.7us`

This is inspired by https://github.com/pytorch/pytorch/commit/a2c17a2b00f7c41866bbde28d33b8c50e5632e01 -- but in Python rather than C++


[ghstack-poisoned]

peterbell10 commented on torch/csrc/dynamo/guards.cpp:
```diff
@@ -585,12 +585,51 @@ static PyObject* assert_size_stride(PyObject* dummy, PyObject* args) {
+static PyObject* _empty_strided_cpu(PyObject* dummy, PyObject* args) {
+  // at::empty_strided is surprising slow.  This is a lower-overhead
+  // version that saves ~2us on every allocation.  Though it is
+  // CPU-only, tuple-only, and only supports 8D tensors.
```
NIT: The 8D part is out of date.

jansel committed:
Update on "[inductor] Slightly faster memory allocation on CPU"


Based on `python benchmarks/dynamo/microbenchmarks/overheads.py`:
- Before `12.2us`
- After `10.5us`

This is inspired by https://github.com/pytorch/pytorch/commit/a2c17a2b00f7c41866bbde28d33b8c50e5632e01 -- but in Python rather than C++


[ghstack-poisoned]

jansel commented:
@pytorchbot merge

jansel committed:
Update on "[inductor] Slightly faster memory allocation on CPU"


Based on `python benchmarks/dynamo/microbenchmarks/overheads.py`:
- Before `12.2us`
- After `10.5us`

This is inspired by https://github.com/pytorch/pytorch/commit/a2c17a2b00f7c41866bbde28d33b8c50e5632e01 -- but in Python rather than C++


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented:
@pytorchbot merge
